invoke-ai/InvokeAI
# Model Manager Refactor Phase 1 - SQL-based config storage

## What type of PR is this? (check all applicable)



## Have you discussed this change with the InvokeAI team?
- [X] Extensively

      
## Have you updated all relevant documentation?


## Description

As discussed with @psychedelicious and @RyanJDick, this is the first phase of the model manager refactor. In this phase, I've added support for storing model configuration information the `invokeai.db` SQL3 database. All the code is separate from the original model manager, so for the time being the frontend is still using the original YAML-based configuration, so the web app still works.

To keep things clean, I've added a new FastAPI route called `model_records` which can add, update, retrieve and delete model records.

The architecture is described in the first section of `docs/contributing/MODEL_MANAGER.md`. 

## QA Instructions, Screenshots, Recordings

There is a pytest for the model sql storage backend in `tests/backend/model_manager_2/test_model_storage_sql.py`.

To populate `invokeai.db` with models from your current `models.yaml`, do the following:

1. Stop the running server
2. Back up `invokeai.db`
3. Run `pip install -e .` to install the command used in the next step.
4. Run `invokeai-migrate-models-to-db`

This will iterate through `models.yaml` and create equivalent database entries in the `model_config` table of `invokeai.db`. Only the models named in the yaml file will be migrated, so anything that is autoloaded will be ignored.

Note that in order to get the `model_records` router to be recognized by the swagger API, I had to rebuild the frontend. Not sure why this was necessary and would appreciate a pointer on a less radical way to do this.

## Added/updated tests?
